File: R/summarywidget.R
#' Show a single summary statistic in a widget
#'
#' A `summarywidget` displays a single statistic derived from a linked table.
#' Its primary use is with the `crosstalk` package. Used with `crosstalk`,
#' a `summarywidget` displays a value which updates as the data selection
#' changes.
#'
#' @param data Data to summarize, normally an instance of [crosstalk::SharedData].
#' @param statistic The statistic to compute.
#' @param column For `sum`, `mean`, `numeric_rate` statistics, the column of `data` to summarize.
#' Not used for `count`, `distinct_count`, `duplicates`, or `count_rate` statistics.
#' @param selection Expression to select a fixed subset of `data`. May be
#' a logical vector or a one-sided formula that evaluates to a logical vector.
#' If used, the `key` given to [crosstalk::SharedData] must be a fixed column (not row numbers).
#' @param numerator_selection Expression to select the numerator subset of `data` for rate calculations.
#' @param denominator_selection Expression to select the denominator subset of `data` for rate calculations.
#' If NULL, the entire data is used as the denominator.
#' @param numerator_column Column to use for the numerator in `numeric_rate` statistic.
#' @param denominator_column Column to use for the denominator in `numeric_rate` statistic.
#' If NULL, `numerator_column` is used.
#' @param digits Number of decimal places to display, or NULL to display full precision.
#' @param big_mark Character used as thousands separator.
#' @param prefix String to prepend to the value. Default is NULL.
#' @param suffix String to append to the value. Default is NULL. Not applied for rate statistics.
#' @param width Width of the widget.
#' @param height Height of the widget.
#' @param elementId Element ID for the widget.
#'
#' @import crosstalk
#' @import htmlwidgets
#'
#' @export
summarywidget <- function(data,
                          statistic = c("count", "sum", "mean", "min", "max", "distinct_count", "duplicates", "count_rate", "numeric_rate"),
                          column = NULL,
                          selection = NULL,
                          numerator_selection = NULL,
                          denominator_selection = NULL,
                          numerator_column = NULL,
                          denominator_column = NULL,
                          digits = 0,
                          big_mark = " ",
                          prefix = NULL,
                          suffix = NULL,
                          width = NULL, height = NULL, elementId = NULL) {

  if (crosstalk::is.SharedData(data)) {
    # Using Crosstalk
    key <- data$key()
    group <- data$groupName()
    data <- data$origData()
  } else {
    # Not using Crosstalk
    warning("summarywidget works best when data is an instance of crosstalk::SharedData.")
    key <- NULL
    group <- NULL
  }

  statistic <- match.arg(statistic)

  # If selection is given, apply it
  if (!is.null(selection)) {
    # Evaluate any formula
    if (inherits(selection, 'formula')) {
      if (length(selection) != 2L)
        stop("Unexpected two-sided formula: ", deparse(selection))
      selection <- eval(selection[[2]], data, environment(selection))
    }

    if (!is.logical(selection))
      stop("Selection must contain TRUE/FALSE values.")
    data <- data[selection, ]
    key <- key[selection]
  }

  # We need to handle numerator and denominator selections
  numerator_flag <- NULL
  denominator_flag <- NULL

  # Prepare the main data value
  if (is.null(column)) {
    if (statistic %in% c('sum', 'mean', 'min', 'max', 'numeric_rate'))
      stop("Column must be provided with ", statistic, " statistic.")
    data_value <- seq_len(nrow(data))  # Use row indices for count-based statistics
  } else {
    if (!(column %in% colnames(data)))
      stop("No ", column, " column in data.")
    data_value <- data[[column]]
  }

  if (statistic %in% c('count_rate', 'numeric_rate')) {

    # Numerator selection
    if (!is.null(numerator_selection)) {
      if (inherits(numerator_selection, 'formula')) {
        if (length(numerator_selection) != 2L)
          stop("Unexpected two-sided formula: ", deparse(numerator_selection))
        numerator_flag <- eval(numerator_selection[[2]], data, environment(numerator_selection))
      } else {
        numerator_flag <- numerator_selection
      }
      if (!is.logical(numerator_flag))
        stop("numerator_selection must contain TRUE/FALSE values.")
    } else {
      stop("numerator_selection must be provided for ", statistic, " statistic.")
    }

    # Denominator selection
    if (!is.null(denominator_selection)) {
      if (inherits(denominator_selection, 'formula')) {
        if (length(denominator_selection) != 2L)
          stop("Unexpected two-sided formula: ", deparse(denominator_selection))
        denominator_flag <- eval(denominator_selection[[2]], data, environment(denominator_selection))
      } else {
        denominator_flag <- denominator_selection
      }
      if (!is.logical(denominator_flag))
        stop("denominator_selection must contain TRUE/FALSE values.")
    } else {
      # Use all data as denominator
      denominator_flag <- rep(TRUE, nrow(data))
    }

    if (statistic == 'numeric_rate') {
      # For 'numeric_rate', we need to specify numerator_column and denominator_column
      if (is.null(numerator_column))
        stop("numerator_column must be provided for numeric_rate statistic.")

      if (!(numerator_column %in% colnames(data)))
        stop("No ", numerator_column, " column in data.")

      if (is.null(denominator_column))
        denominator_column <- numerator_column  # Use same column if denominator_column not provided

      if (!(denominator_column %in% colnames(data)))
        stop("No ", denominator_column, " column in data.")

      data$numerator_value <- data[[numerator_column]]
      data$denominator_value <- data[[denominator_column]]
    } else {
      data$numerator_value <- NULL
      data$denominator_value <- NULL
    }
  }

  # Prepare data to pass to JavaScript
  x <- list(
    data = data_value,
    key = key,
    numerator_flag = numerator_flag,
    denominator_flag = denominator_flag,
    numerator_value = if (!is.null(data$numerator_value)) data$numerator_value else NULL,
    denominator_value = if (!is.null(data$denominator_value)) data$denominator_value else NULL,
    settings = list(
      statistic = statistic,
      digits = digits,
      big_mark = big_mark,
      prefix = prefix,
      suffix = suffix,
      crosstalk_group = group
    )
  )

  # Create widget
  htmlwidgets::createWidget(
    name = 'summarywidget',
    x,
    width = width,
    height = height,
    package = 'summarywidget',
    elementId = elementId,
    dependencies = crosstalk::crosstalkLibs()
  )
}


#' Shiny bindings for summarywidget
#'
#' Output and render functions for using summarywidget within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a summarywidget
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name summarywidget-shiny
#'
#' @export
summarywidgetOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'summarywidget', width, height, package = 'summarywidget')
}

#' @rdname summarywidget-shiny
#' @export
renderSummarywidget <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, summarywidgetOutput, env, quoted = TRUE)
}

# Use a <span> container rather than the default <div>
summarywidget_html <- function(id, style, class, ...){
  htmltools::tags$span(id = id, class = class)
}

File: docs/site_libs/summarywidget-binding-0.0.0.9000/summarywidget.js
HTMLWidgets.widget({
  name: 'summarywidget',
  type: 'output',

  factory: function(el, width, height) {

    // Function to filter an object based on keys
    var filterKeys = function(obj, keys) {
      var result = {};
      keys.forEach(function(k) {
        if (obj.hasOwnProperty(k))
          result[k] = obj[k];
      });
      return result;
    };

    return {
      renderValue: function(x) {

        // Prepare data object with keys
        var data = {};
        var i;
        for (i = 0; i < x.data.length; i++) {
          var key = x.key ? x.key[i] : i;
          data[key] = {
            value: x.data[i],
            numerator_flag: x.numerator_flag ? x.numerator_flag[i] === true : false,
            denominator_flag: x.denominator_flag ? x.denominator_flag[i] === true : false,
            numerator_value: x.numerator_value ? x.numerator_value[i] : null,
            denominator_value: x.denominator_value ? x.denominator_value[i] : null
          };
        }

        // Function to update the display based on filtered data
        var update = function(d) {
          // Convert object to array
          var dataArray = [];
          for (var key in d) {
            if (d.hasOwnProperty(key)) {
              dataArray.push(d[key]);
            }
          }

          var value = 0; // Initialize the value to display

          // Compute the statistic based on the selected type
          switch (x.settings.statistic) {
            case 'count_rate':
              // Filter numerator and denominator data
              var numeratorData = dataArray.filter(function(d) { return d.numerator_flag; });
              var denominatorData = dataArray.filter(function(d) { return d.denominator_flag; });
              var numeratorCount = numeratorData.length;
              var denominatorCount = denominatorData.length;
              value = denominatorCount === 0 ? 0 : (numeratorCount / denominatorCount * 100);
              break;

            case 'numeric_rate':
              // Filter numerator and denominator data
              var numeratorData = dataArray.filter(function(d) { return d.numerator_flag; });
              var denominatorData = dataArray.filter(function(d) { return d.denominator_flag; });
              // Sum numerator and denominator values
              var numeratorSum = numeratorData.reduce(function(acc, d) {
                return acc + (d.numerator_value || 0);
              }, 0);
              var denominatorSum = denominatorData.reduce(function(acc, d) {
                return acc + (d.denominator_value || 0);
              }, 0);
              value = denominatorSum === 0 ? 0 : (numeratorSum / denominatorSum * 100);
              break;

            case 'count':
              value = dataArray.length;
              break;

            case 'sum':
              value = dataArray.reduce(function(acc, d) { return acc + d.value; }, 0);
              break;

            case 'mean':
              value = dataArray.reduce(function(acc, d) { return acc + d.value; }, 0) / dataArray.length;
              break;

            case 'distinct_count':
              var uniqueValues = {};
              dataArray.forEach(function(d) { uniqueValues[d.value] = true; });
              value = Object.keys(uniqueValues).length;
              break;

            case 'duplicates':
              var counts = {};
              dataArray.forEach(function(d) {
                counts[d.value] = (counts[d.value] || 0) + 1;
              });
              value = Object.values(counts).filter(function(count) { return count > 1; }).length;
              break;

            case 'min':
              value = Math.min.apply(null, dataArray.map(function(d) { return d.value; }));
              break;

            case 'max':
              value = Math.max.apply(null, dataArray.map(function(d) { return d.value; }));
              break;

            default:
              console.error('Invalid statistic specified:', x.settings.statistic);
              return;
          }

          // Function to format numbers with thousand separators
          function numberWithSep(x, bigMark) {
            if (typeof x !== "string") {
              x = x.toString();
            }
            var parts = x.split(".");
            parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, bigMark);
            return parts.join(".");
          }

          // Apply formatting
          if (x.settings.digits !== null) {
            value = parseFloat(value).toFixed(x.settings.digits);
          }
          if (x.settings.big_mark) {
            value = numberWithSep(value, x.settings.big_mark);
          }

          // Apply prefix and suffix
          var displayValue = value;

          // Always apply prefix if provided
          if (x.settings.prefix) {
            displayValue = x.settings.prefix + displayValue;
          }

          // For rate statistics, append '%' and skip suffix
          if (x.settings.statistic === 'count_rate' || x.settings.statistic === 'numeric_rate') {
            displayValue = displayValue + '%';
            // Do not apply suffix
          } else {
            // Apply suffix if provided
            if (x.settings.suffix) {
              displayValue = displayValue + x.settings.suffix;
            }
          }

          // Update the element's text content
          el.innerText = displayValue;
        };

        // Set up Crosstalk filter and selection handles
        var ct_filter = new crosstalk.FilterHandle();
        ct_filter.setGroup(x.settings.crosstalk_group);
        ct_filter.on("change", function(e) {
          if (e.value) {
            update(filterKeys(data, e.value));
          } else {
            update(data);
          }
        });

        var ct_sel = new crosstalk.SelectionHandle();
        ct_sel.setGroup(x.settings.crosstalk_group);
        ct_sel.on("change", function(e) {
          if (e.value && e.value.length) {
            update(filterKeys(data, e.value));
          } else {
            update(data);
          }
        });

        // Initial update
        update(data);
      },

      resize: function(width, height) {
        // Code to handle resizing if necessary
        // Currently not needed as the widget displays a single value
      }

    };
  }
});
